cpp/src/request_mem.cpp

```diff
@@ -0,0 +1,196 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <cstdio>
+#include <memory>
+#include <string>
+
+#include <ucp/api/ucp.h>
+
+#include <ucxx/buffer.h>
+#include <ucxx/delayed_submission.h>
+#include <ucxx/request_mem.h>
+
+namespace ucxx {
+
+std::shared_ptr<RequestMem> createRequestMem(
+  std::shared_ptr<Endpoint> endpoint,
+  const std::variant<data::MemPut, data::MemGet> requestData,
+  const bool enablePythonFuture                = false,
+  RequestCallbackUserFunction callbackFunction = nullptr,
+  RequestCallbackUserData callbackData         = nullptr)
+{
+  std::shared_ptr<RequestMem> req = std::visit(
+    data::dispatch{
+      [&endpoint, &enablePythonFuture, &callbackFunction, &callbackData](data::MemPut memPut) {
+        return std::shared_ptr<RequestMem>(new RequestMem(
+          endpoint, memPut, "memPut", enablePythonFuture, callbackFunction, callbackData));
+      },
+      [&endpoint, &enablePythonFuture, &callbackFunction, &callbackData](data::MemGet memGet) {
+        return std::shared_ptr<RequestMem>(new RequestMem(
+          endpoint, memGet, "memGet", enablePythonFuture, callbackFunction, callbackData));
+      },
+    },
+    requestData);
```

Comment:
Because you are doing a `std::visit` inside of the constructor, couldnt this be simplified to just:
```cpp
std::shared_ptr<RequestMem> req = std::shared_ptr<RequestMem>(new RequestMem(
          endpoint, requestData, isPut ? "memPut" : "memGet", enablePythonFuture, callbackFunction, callbackData));
```
And even the `isPut ? "memPut" : "memGet"`` could be removed and set in the dispatch inside of the constructor since those are the only two options. Seems a bit redundant to have a `std::visit` followed by another `std::visit`

Reply:
From a simplistic point-of-view, you're right. However, what we're attempting to solve with `std::variant`/`std::visit` is to provide strong-typing, and thus prevent things like the caller from mistakenly shuffling arguments. I agree this adds some overhead that is unnecessary from a functional standpoint, but a "safe" interface is generally preferable. An alternative would be adding a type for `isPut` that is not a `bool` which would be fine for this two-case scenario, but that is another type of complexity we would need to add so there would be a tradeoff as well that I feel would be more confusing, even though it could slightly decrease code length.